da/da_test.go

```diff
@@ -0,0 +1,124 @@
+package da
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/rollkit/go-da/proxy"
+	goDATest "github.com/rollkit/go-da/test"
+	"github.com/rollkit/rollkit/types"
+)
+
+const mockDaBlockTime = 100 * time.Millisecond
+
+func TestMain(m *testing.M) {
+	srv := startMockGRPCServ()
+	if srv == nil {
+		os.Exit(1)
+	}
+	exitCode := m.Run()
+
+	// teardown servers
+	srv.GracefulStop()
+
+	os.Exit(exitCode)
+}
+
+func TestRetrieve(t *testing.T) {
+	dummyClient := &DAClient{DA: goDATest.NewDummyDA(), Logger: log.TestingLogger()}
+	grpcClient, err := startMockGRPCClient()
+	require.NoError(t, err)
+	clients := map[string]*DAClient{
+		"dummy": dummyClient,
+		"grpc":  grpcClient,
+	}
+	for name, dalc := range clients {
+		t.Run(name, func(t *testing.T) {
+			doTestRetrieve(t, dalc)
+		})
+	}
+}
+
+func startMockGRPCServ() *grpc.Server {
+	srv := proxy.NewServer(goDATest.NewDummyDA(), grpc.Creds(insecure.NewCredentials()))
+	lis, err := net.Listen("tcp", "127.0.0.1"+":"+strconv.Itoa(7980))
+	if err != nil {
+		fmt.Println(err)
+		return nil
+	}
+	go func() {
+		_ = srv.Serve(lis)
+	}()
+	return srv
+}
+
+func startMockGRPCClient() (*DAClient, error) {
+	client := proxy.NewClient()
+	err := client.Start("127.0.0.1:7980", grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	return &DAClient{DA: client, Logger: log.TestingLogger()}, nil
+}
+
+func doTestRetrieve(t *testing.T, dalc *DAClient) {
```

Comment:
We don't have a test like `TestSubmitBlocks` so is this testing both `SubmitBlocks` and `RetrieveBlocks`? If so, we should either rename this test to include that, or add a separate test for `SubmitBlocks`.